pkg/ocicni/ocicni.go

```diff
@@ -511,6 +491,74 @@ func (network *cniNetwork) addToNetwork(cacheDir string, podNetwork *PodNetwork,
 	return res, nil
 }
 
+func (network *cniNetwork) checkNetwork(cacheDir string, podNetwork *PodNetwork, ifName string, runtimeConfig RuntimeConfig, nsManager *nsManager) (cnitypes.Result, error) {
+
+	rt, err := buildCNIRuntimeConf(cacheDir, podNetwork, ifName, runtimeConfig)
+	if err != nil {
+		logrus.Errorf("Error checking network: %v", err)
+		return nil, err
+	}
+
+	netconf, cninet := network.NetworkConfig, network.CNIConfig
+	logrus.Infof("About to check CNI network %s (type=%v)", netconf.Name, netconf.Plugins[0].Network.Type)
+
+	gtet, err := cniversion.GreaterThanOrEqualTo(netconf.CNIVersion, "0.4.0")
+	if err != nil {
+		return nil, err
+	}
+
+	var result cnitypes.Result
+
+	// When CNIVersion supports Check, use it.  Otherwise fall back on what was done initially.
+	if gtet {
+		err = cninet.CheckNetworkList(context.Background(), netconf, rt)
+		logrus.Infof("Checking CNI network %s (config version=%v)", netconf.Name, netconf.CNIVersion)
+		if err != nil {
+			logrus.Errorf("Error checking network: %v", err)
+			return nil, err
+		}
+		result, err = cninet.GetNetworkListCachedResult(netconf, rt)
+		if err != nil {
+			logrus.Errorf("Error GetNetworkListCachedResult: %v", err)
+			return nil, err
+		}
+
+		return result, nil
```

Comment:
@mccv1r0 it's possible that the cached result won't exist, like if CRIO is upgraded while pods are running. So we can get (result==nil && err==nil). But also, libcni will cache the result regardless of the result's CNI version. So we can always use the cached result codepath. We just can't use the CheckNetworkList() codepath for < 0.4.0.

So I think this should actually be something more like:

```
if gtet {
    err = cninet.CheckNetworkList()
    if err != nil {
        ...
    }
    <fall through to get cached result>
}

result, err == cninet.GetNetworkListCachedResult(netconf, rt)
if err != nil {
    ...
} else if result != nil {
    return result, nli
}

<do old fallback stuff>
```

Reply:
You're suggesting to always use result from GetNetworkListCachedResult() regardless of CNI Version... and only resort to original code when those results are nil (and even in the >= 0.4.0 case)?